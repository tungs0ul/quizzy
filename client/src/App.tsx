import { useState, useEffect } from "react";
import useWebSocket from "react-use-websocket";
import useLocalStorage from "./hooks/useLocalStorage";
import StudentView from "./components/views/StudentView";
import AdminView from "./components/views/AdminView";
import { BrowserRouter, Routes, Route, Link } from "react-router-dom";

// const WSURI = "ws://localhost:8080/ws";
const WSURI =
  ((window.location.protocol == "https:" && "wss://") || "ws://") +
  window.location.host +
  "/ws";

const initQuestion = {
  msg: "question",
  data: "",
  options: [],
};

export type Question = {
  type: string;
  data: string;
  options: string[];
};

export type Result = {
  name: string;
  value: number;
};

function App() {
  const { sendMessage, lastMessage } = useWebSocket(WSURI, {
    shouldReconnect: () => true,
    reconnectInterval: 5000,
  });
  const [question, setQuestion] = useLocalStorage("question", initQuestion);
  const [answer, setAnswer] = useLocalStorage("answer", null);
  const [result, setResult] = useState<Result[]>([]);
  const [notice, setNotice] = useState("");
  const [active, setActive] = useState(false);
  const [intervalID, setIntervalID] = useState<number | null>(null);

  useEffect(() => {
    if (lastMessage !== null) {
      if (!lastMessage.data) {
        return;
      }
      const data = JSON.parse(lastMessage.data);
      if (data.type === "question") {
        if (data.data !== question.data) {
          setNotice("The question has been changed!");
          setAnswer(null);
          setQuestion(data);
          setResult([]);
          if (intervalID !== null) {
            clearInterval(intervalID);
          }
          let id = setInterval(() => setNotice(""), 1500);
          setIntervalID(id);
        }
      } else if (data.type === "result") {
        if (
          data.data.reduce((acc: number, cur: number) => acc + cur, 0) === 0
        ) {
          setAnswer(null);
        }
        let result: Result[] = [];
        data.data.forEach((e: number, i: number) => {
          result.push({ name: question.options[i], value: e });
        });
        setResult(result);
      } else if (data.type === "clear") {
        setAnswer(null);
        setNotice("The answer has been cleared!");
        setInterval(() => setNotice(""), 1500);
      } else if (data.type === "reset") {
        setQuestion(initQuestion);
        setAnswer(null);
        setResult([]);
      }
    }
  }, [lastMessage]);

  return (
    <BrowserRouter>
      <div className="h-screen relative">
        <div className="flex justify-between items-center pb-4 pr-10 header">
          <div className="cursor-pointer p-4 text-yellow-600 hover:text-green-600 text-4xl items-center h-full flex justify-center">
            <Link to="/login">Admin</Link>
          </div>
          <div className="flex justify-end tablet:text-12xl laptop:text-16xl mobile:text-6xl font-bold">
            QUIZZY
          </div>
        </div>
        <Routes>
          <Route
            path="/"
            element={
              <StudentView
                answer={answer}
                question={question}
                result={result}
                sendMessage={sendMessage}
                setAnswer={setAnswer}
                notice={notice}
              />
            }
          />
          <Route
            path="/login"
            element={
              <AdminView
                sendMessage={sendMessage}
                setResult={setResult}
                result={result}
                question={question}
              />
            }
          />
        </Routes>
      </div>
    </BrowserRouter>
  );
}

export default App;
