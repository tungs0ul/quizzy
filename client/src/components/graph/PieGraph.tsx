import React, { useEffect, useState } from "react";
import { PieChart, Pie, Cell, Tooltip, Legend } from "recharts";
import { COLORS } from "../../utils";
import { Result } from "../../App";

const RADIAN = Math.PI / 180;

const renderCustomizedLabel = ({
  cx,
  cy,
  midAngle,
  innerRadius,
  outerRadius,
  percent,
  _,
}: any) => {
  const radius = innerRadius + (outerRadius - innerRadius) * 0.5;
  const x = cx + radius * Math.cos(-midAngle * RADIAN);
  const y = cy + radius * Math.sin(-midAngle * RADIAN);

  if (!percent) {
    return <></>;
  }

  return (
    <text
      x={x}
      y={y}
      fill="white"
      textAnchor={x > cx ? "start" : "end"}
      dominantBaseline="central"
    >
      {`${(percent * 100).toFixed(0)}%`}
    </text>
  );
};

type Props = {
  data: Result[];
  resize?: boolean;
};

export default function PieGraph({ data, resize }: Props) {
  const [width, setWidth] = useState(
    resize
      ? Math.min(window.innerWidth / 2, window.innerHeight) * 0.75
      : Math.min(window.innerWidth, window.innerHeight) * 0.75
  );

  useEffect(() => {
    const handleResize = () => {
      let size = Math.min(window.innerWidth, window.innerHeight) * 0.75;
      if (resize) {
        size = size * 0.5;
      }
      setWidth(size);
    };
    window.addEventListener("resize", handleResize);
    return () => window.removeEventListener("resize", handleResize);
  });

  return (
    <div className="flex justify-center cursor-pointer">
      <PieChart width={width} height={width}>
        <Legend layout="horizontal" verticalAlign="top" align="center" />
        <Tooltip />
        <Pie
          data={data}
          cx="50%"
          cy="50%"
          labelLine={false}
          label={renderCustomizedLabel}
          outerRadius={width / 2 - 50}
          fill="#8884d8"
          dataKey="value"
        >
          {data.map((_: any, index: number) => (
            <Cell key={`cell-${index}`} fill={COLORS[index % COLORS.length]} />
          ))}
        </Pie>
      </PieChart>
    </div>
  );
}
