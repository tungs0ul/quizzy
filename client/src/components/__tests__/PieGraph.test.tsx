import PieGraph from "../graph/PieGraph";
import TestRenderer, { act } from "react-test-renderer";
import { Result } from "../../App";
import * as React from "react";

const data = [
  { name: "1", value: 1 },
  { name: "2", value: 1 },
];

it("render piegraph", () => {
  const test = TestRenderer.create(<PieGraph data={data} resize={true} />);
  act(() => {
    global.innerWidth = 500;
    global.dispatchEvent(new Event("resize"));
    expect(test.toJSON()).toMatchSnapshot();
  });
});
