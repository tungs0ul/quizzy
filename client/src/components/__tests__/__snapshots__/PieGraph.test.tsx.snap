// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render piegraph 1`] = `
<div
  className="flex justify-center cursor-pointer"
>
  <div
    className="recharts-wrapper"
    style={
      Object {
        "cursor": "default",
        "height": 460.8,
        "position": "relative",
        "width": 460.8,
      }
    }
  >
    <svg
      className="recharts-surface"
      cx="50%"
      cy="50%"
      height={460.8}
      version="1.1"
      viewBox="0 0 460.8 460.8"
      width={460.8}
    >
      <defs>
        <clipPath
          id="recharts1-clip"
        >
          <rect
            height={450.8}
            width={450.8}
            x={5}
            y={5}
          />
        </clipPath>
      </defs>
      <g
        className="recharts-layer recharts-pie"
      >
        <g
          className="recharts-layer"
        >
          <g
            className="recharts-layer recharts-pie-sector"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          />
          <g
            className="recharts-layer recharts-pie-sector"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          />
        </g>
      </g>
    </svg>
    <div
      className="recharts-legend-wrapper"
      style={
        Object {
          "height": "auto",
          "left": 5,
          "position": "absolute",
          "top": 5,
          "width": 450.8,
        }
      }
    >
      <ul
        className="recharts-default-legend"
        style={
          Object {
            "margin": 0,
            "padding": 0,
            "textAlign": "center",
          }
        }
      >
        <li
          className="recharts-legend-item legend-item-0"
          style={
            Object {
              "display": "inline-block",
              "marginRight": 10,
            }
          }
        >
          <svg
            className="recharts-surface"
            height={14}
            style={
              Object {
                "display": "inline-block",
                "marginRight": 4,
                "verticalAlign": "middle",
              }
            }
            version="1.1"
            viewBox="0 0 32 32"
            width={14}
          >
            <path
              className="recharts-legend-icon"
              d="M0,4h32v24h-32z"
              fill="#0088FE"
              stroke="none"
            />
          </svg>
          <span
            className="recharts-legend-item-text"
            style={
              Object {
                "color": "#0088FE",
              }
            }
          >
            1
          </span>
        </li>
        <li
          className="recharts-legend-item legend-item-1"
          style={
            Object {
              "display": "inline-block",
              "marginRight": 10,
            }
          }
        >
          <svg
            className="recharts-surface"
            height={14}
            style={
              Object {
                "display": "inline-block",
                "marginRight": 4,
                "verticalAlign": "middle",
              }
            }
            version="1.1"
            viewBox="0 0 32 32"
            width={14}
          >
            <path
              className="recharts-legend-icon"
              d="M0,4h32v24h-32z"
              fill="#00C49F"
              stroke="none"
            />
          </svg>
          <span
            className="recharts-legend-item-text"
            style={
              Object {
                "color": "#00C49F",
              }
            }
          >
            2
          </span>
        </li>
      </ul>
    </div>
    <div
      className="recharts-tooltip-wrapper"
      style={
        Object {
          "MozTransform": "translate(undefinedpx, undefinedpx)",
          "OTransform": "translate(undefinedpx, undefinedpx)",
          "WebkitTransform": "translate(undefinedpx, undefinedpx)",
          "left": 0,
          "msTransform": "translate(undefinedpx, undefinedpx)",
          "pointerEvents": "none",
          "position": "absolute",
          "top": 0,
          "transform": "translate(undefinedpx, undefinedpx)",
          "visibility": "hidden",
        }
      }
    >
      <div
        className="recharts-default-tooltip"
        style={
          Object {
            "backgroundColor": "#fff",
            "border": "1px solid #ccc",
            "margin": 0,
            "padding": 10,
            "whiteSpace": "nowrap",
          }
        }
      >
        <p
          className="recharts-tooltip-label"
          style={
            Object {
              "margin": 0,
            }
          }
        >
          
        </p>
      </div>
    </div>
  </div>
</div>
`;
